File: ws/src/cpp_pkg/src/turtle_spawner_node.cpp
#include "cpp_pkg/turtle_spawner_node.hpp"

TurtleSpawnerNode::TurtleSpawnerNode()
    : Node("turtle_spawner_node")
{
    this->declare_parameter("spawn_frequency", 1);
    spawn_frequency_ = this->get_parameter("spawn_frequency").as_int();
    spawn_timer_ = this->create_wall_timer(std::chrono::seconds(1/spawn_frequency_), std::bind(&TurtleSpawnerNode::spawn_timer_callback, this));
    turtles_pub_ = this->create_publisher<interfaces::msg::TurtleArray>("/alive_turtles", 10);
    catch_service_ = this->create_service<interfaces::srv::CatchTurtle>("/catch_turtle", std::bind(&TurtleSpawnerNode::catch_service_callback, this, _1, _2));
    // TODO: GET NAME PARAMETER
    RCLCPP_INFO(this->get_logger(), "turtle_spawner_node started ...");
}

void TurtleSpawnerNode::catch_service_callback(interfaces::srv::CatchTurtle::Request::SharedPtr request, interfaces::srv::CatchTurtle::Response::SharedPtr response)
{
    client_threads_.push_back(std::thread(std::bind(&TurtleSpawnerNode::call_kill_service, this, request->name)));
    response = std::make_shared<interfaces::srv::CatchTurtle::Response>();
    return;
}

void TurtleSpawnerNode::call_kill_service(const std::string &name)
{
    auto kill_client = create_client<turtlesim::srv::Kill>("/kill");
    while (!kill_client->wait_for_service(1s))
    {
        RCLCPP_WARN(this->get_logger(), "Waiting for /kill service ...");
    }
    auto request = std::make_shared<turtlesim::srv::Kill::Request>();
    request->name = name;
    auto future = kill_client->async_send_request(request);

    try
    {
        auto response = future.get();
        for (std::vector<interfaces::msg::Turtle>::iterator itr = alive_turtles_.begin(); itr < alive_turtles_.end(); itr++)
        {
            if (itr->name == request->name)
            {
                alive_turtles_.erase(itr);
                break;
            }
        }
        RCLCPP_WARN(this->get_logger(), std::string(std::string("Killed ") + name).c_str());
    }
    catch (const std::exception &e)
    {
        RCLCPP_ERROR(this->get_logger(), e.what());
    }
    return;
}

void TurtleSpawnerNode::spawn_timer_callback()
{
    client_threads_.push_back(std::thread(std::bind(&TurtleSpawnerNode::call_spawn_service, this)));
    return;
}

void TurtleSpawnerNode::call_spawn_service()
{
    auto spawn_client = this->create_client<turtlesim::srv::Spawn>("/spawn");
    while (!spawn_client->wait_for_service(1s))
    {
        RCLCPP_WARN(this->get_logger(), "Waiting for /spawn service ...");
    }
    static std::default_random_engine engine;
    std::uniform_real_distribution<float> position(0.5, 10.5);
    std::uniform_real_distribution<float> angle(0.0, 2 * M_PI);
    auto request = std::make_shared<turtlesim::srv::Spawn::Request>();
    request->x = position(engine);
    request->y = position(engine);
    request->theta = angle(engine);
    auto future = spawn_client->async_send_request(request);
    try
    {
        auto response = future.get();
        interfaces::msg::Turtle new_turtle;
        new_turtle.pose[0] = request->x;
        new_turtle.pose[1] = request->y;
        new_turtle.pose[2] = request->theta;
        new_turtle.name = response->name;
        alive_turtles_.push_back(new_turtle);
        interfaces::msg::TurtleArray turtles;
        turtles.alive_turtles = alive_turtles_;
        turtles_pub_->publish(turtles);
        RCLCPP_INFO(this->get_logger(), std::string(std::string("Spawned ") + response->name).c_str());
    }
    catch (const std::exception &e)
    {
        RCLCPP_ERROR(this->get_logger(), e.what());
    }
    return;
}

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TurtleSpawnerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ws/src/python_pkg/python_pkg/turtle_spawner_node.py
import rclpy
from random import uniform
from math import pi
from functools import partial
from rclpy.node import Node
from turtlesim.srv import Spawn
from interfaces.msg import Turtle, TurtleArray
from interfaces.srv import CatchTurtle
from turtlesim.srv import Kill

class TurtleSpawnerNode(Node):
    def __init__(self):
        super().__init__("turtle_spawner_node")
        self.declare_parameter("spawn_frequency", 1)
        self.spawn_frequency = self.get_parameter("spawn_frequency").get_parameter_value().integer_value
        self.spawn_timer = self.create_timer(1/self.spawn_frequency, self.call_spawn_service)

        self.turtle_pub = self.create_publisher(TurtleArray, "/alive_turtles", 10)
        self.alive_turtles = []

        self.catch_service = self.create_service(CatchTurtle, "/catch_turtle", self.call_kill_service)

        self.get_logger().info("turtle_spawner_node started...")


    def call_kill_service(self, request, response):
        kill_client = self.create_client(Kill, "/kill")
        while not kill_client.wait_for_service(1.0):
            self.get_logger().warn("waiting for /kill service...")
        kill_request = Kill.Request()
        kill_request.name = request.name
        future = kill_client.call_async(kill_request)
        future.add_done_callback(partial(self.kill_service_callback, name=request.name))
        return response

    def kill_service_callback(self, future, name):
        try:
            response = future.result()
            for turtle in self.alive_turtles:
                if turtle.name == name:
                    self.alive_turtles.remove(turtle)
            self.get_logger().info("Killed {}".format(name))
        except Exception as e:
            self.get_logger().error("Service call failed: {}".format(e))

    def call_spawn_service(self):
        spawn_client = self.create_client(Spawn, "/spawn")
        while not spawn_client.wait_for_service(1.0):
            self.get_logger().warn("waiting for /spawn service...")
        request = Spawn.Request()
        request.x = uniform(0.5, 10.5)
        request.y = uniform(0.5, 10.5)
        request.theta = uniform(0, 2*pi)
        future = spawn_client.call_async(request)
        future.add_done_callback(partial(self.spawn_service_callback, x=request.x, y=request.y, theta=request.theta))

    def spawn_service_callback(self, future, x, y, theta):
        try:
            response = future.result()
            self.get_logger().info("Spawned {}".format(response.name))
            new_turtle = Turtle()
            new_turtle.name = response.name
            new_turtle.pose[0] = x
            new_turtle.pose[1] = y
            new_turtle.pose[2] = theta
            self.alive_turtles.append(new_turtle)

            turtle_array = TurtleArray()
            turtle_array.alive_turtles = self.alive_turtles
            self.turtle_pub.publish(turtle_array)
        except Exception as e:
            self.get_logger().error("Service call failed: {}".format(e))


def main(args=None):
    rclpy.init(args=args)
    node = TurtleSpawnerNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()